### ISSUE TITLE
Missing: Accessibility

### ISSUE THREAD

**Issue by migueldeicaza** [migueldeicaza/SwiftTerm#12]

**Comment by holzschu** [comment]
Hi, 
I'm trying to work on this issue. As you know, iOS has a good screen reader (VoiceOver). When it works well, it allows users to tap anywhere on the screen and have the system read to them the line at that point on screen.

In theory, it works by setting `accessibilityValue` to the screen content, and with a `UITextView` it's supposed to be almost automatic. In practice, I haven't been able to make it work with SwiftTerm. I can set `accessibilityLabel` to something, and that something will be read by VoiceOver, but setting `accessibilityValue` does not do anything. In some tests, it feels like something else is overriding what I put in `accessibilityValue`, to put information about `SwiftTerm` being a text field, currently editing, but nothing about the actual content.

Would you have any pointers where I should start looking?

**Comment by holzschu** [comment]
Hi again, answering my own question: adding the following lines to `iOSTerminalView.swift` provides a working screen reader:
```swift
extension TerminalView: UIAccessibilityReadingContent {
    public func accessibilityLineNumber(for point: CGPoint) -> Int {
        return Int(floor(max(point.y,0) / cellDimension.height))
        return 0
    }
    
    public func accessibilityContent(forLineNumber lineNumber: Int) -> String? {
        let start = Position(col: 0, row: lineNumber)
        let end = Position(col: terminal.buffer.lines[lineNumber].count, row: lineNumber)
        return terminal.getDisplayText(start: start, end: end)
    }
    
    public func accessibilityFrame(forLineNumber lineNumber: Int) -> CGRect {
        let topVisibleLine = Int(contentOffset.y/cellDimension.height)
        let lineOffset =  cellDimension.height * CGFloat (lineNumber - topVisibleLine + 1)
        let lineOrigin = CGPoint(x: 0, y: lineOffset)
        let columnCount = terminal.buffer.lines[lineNumber].count
        var rect = CGRect(
            x: lineOrigin.x,
            y: lineOrigin.y + 3,
            width: CGFloat(columnCount) * cellDimension.width, height: cellDimension.height)
        return rect
    }
    
    public func accessibilityPageContent() -> String? {
        // Use full screen for accessibility:
        let start = Position(col: 0, row: 0)
        let lastLine = terminal.buffer.lines.count - 1
        let end = Position(col: terminal.buffer.lines[lastLine].count, row: lastLine)
        return terminal.getDisplayText(start: start, end: end)
    }
}
```
Reference: https://developer.apple.com/videos/play/wwdc2019/248

**Comment by migueldeicaza** [comment]
Oh I love this!   There were some hooks in xterm.js that existed when I first did the port that might be useful here too, something on scroll or something like that.

**Comment by holzschu** [comment]
Yes, that would be `UIAccessibilityScrollDirection.next` and `UIAccessibilityScrollDirection.previous`. In my tests, the terminal window scrolls fine without them (three-fingers swipe up and down) but I'm waiting for the users to comment. 

I would be more interested in those functions:
```swift 
func accessibilityAttributedContent(forLineNumber: Int) -> NSAttributedString?
func accessibilityAttributedPageContent() -> NSAttributedString?
```
that return attributed strings for a given line or the entire page (instead of just the text content). But currently there's only `getText()`, not `getAttributedText()`.

**Comment by migueldeicaza** [comment]
Let me look at that, this sounds like a good idea.

**Comment by migueldeicaza** [comment]
I added those suggestions to accessible-attributed-string can you take it for a spin?

**Comment by holzschu** [comment]
So the result of my tests is that it "works", in the sense that the attributed-string versions are called instead of the standard versions, but it doesn't make any audible difference (VoiceOver does not read bold, italic text or colored text in a different way).

A markdown line with `**test** *italic*` is returned as the following attributed string:
```
{
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedSRGBColorSpace 0.917647 0.92549 0.137255 1";
    NSFont = "<UICTFont: 0x12fd25d50> font-family: \"Menlo-Regular\"; font-weight: normal; font-style: normal; font-size: 13.00pt";
}test{
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedSRGBColorSpace 0 1 0 1";
    NSFont = "<UICTFont: 0x12fd372e0> font-family: \"Menlo\"; font-weight: bold; font-style: normal; font-size: 13.00pt";
}  {
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedSRGBColorSpace 0 1 0 1";
    NSFont = "<UICTFont: 0x12fd25d50> font-family: \"Menlo-Regular\"; font-weight: normal; font-style: normal; font-size: 13.00pt";
}italic{
    NSBackgroundColor = "UIExtendedGrayColorSpace 0 0";
    NSColor = "UIExtendedSRGBColorSpace 0 1 0 1";
    NSFont = "<UICTFont: 0x12fd37510> font-family: \"Menlo\"; font-weight: normal; font-style: italic; font-size: 13.00pt";
}
```  
~~I'm not an expert on attributed string, but the attributes seem to be reversed: they appear after the text they should be applied to.~~ which is the normal encoding for "test" in bold and "italic" in italic, so it just seems that VoiceOver does not read bold or italic text in a different way.

In the WWDC presentation, they used AttributedText to tell VoiceOver that part of the string is in a different language, which does not aply here.

**Comment by holzschu** [comment]
On the `accessibilityScroll` function, here's what I found:
- it currently is not called for a single-finger swipe, only for a three-fingers swipe (users tell me single-finger swipe should work) (puzzling),
- you needs cases for `.down`, `.right`, `.up` and `.left`, not just `.next` and `.previous` (an easy fix),
- once it's called, it reads the entire app starting from the first line, while I think the expected behavior is reading the current page (a relatively easy fix).

On a separate note, I've tried setting the `accessibilityContent(forLineNumber lineNumber: Int)` function so that it extends to the previous and next line if it's a wrapped line, using `isWrapped`. It seems that after a certain amount of scrolling (around 500 lines?), the value for `isWrapped` becomes different even though the line hasn't changed.
